pkg/guacanalytics/toposort.go

```diff
@@ -0,0 +1,66 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"fmt"
+)
+
+// TODO: add tests
+func ToposortFromBfsNodeMap(nodeMap map[string]BfsNode) (map[int][]string, error) {
+	frontiers := make(map[int][]string)
+	parentsMap := copyParents(nodeMap)
+	frontierLevel := 0
+	numNodes := 0
+	totalNodes := len(parentsMap)
+
+	for numNodes <= totalNodes {
+		foundIDs := make(map[string]bool)
+		for id, parentsList := range parentsMap {
+			if len(parentsList) == 0 {
+				frontiers[frontierLevel] = append(frontiers[frontierLevel], id)
+				foundIDs[id] = true
+				numNodes++
+				delete(parentsMap, id)
+			}
+		}
+
+		if len(foundIDs) == 0 {
+			return frontiers, fmt.Errorf("Error: cycle detected")
+		}
+
+		for id, parentsList := range parentsMap {
+			for index, parentID := range parentsList {
+				if foundIDs[parentID] {
+					parentsMap[id] = append(parentsList[:index], parentsList[index+1:]...)
```

Comment:
what is the behavior of modifying the list that is being iterated on?

Reply:
Good point-- it could cause some sort of exception.  I'll change it to copy over the list before iterating over it.